kvdb/sqlite/db.go

```diff
@@ -0,0 +1,207 @@
+	_ "modernc.org/sqlite" // Register relevant drivers.
+)
+
+const (
+	// sqliteOptionPrefix is the string prefix sqlite uses to set various
+	// options. This is used in the following format:
+	//   * sqliteOptionPrefix || option_name = option_value.
+	sqliteOptionPrefix = "_pragma"
+
+	// sqliteTxLockImmediate is a dsn option used to ensure that write
+	// transactions are started immediately.
+	sqliteTxLockImmediate = "_txlock=immediate"
+)
+
+// NewSqliteBackend returns a db object initialized with the passed backend
+// config. If a sqlite connection cannot be established, then an error is
+// returned.
+func NewSqliteBackend(ctx context.Context, cfg *Config, dbPath, fileName,
+	prefix string) (walletdb.DB, error) {
+
+	// First, we add a set of mandatory pragma options to the query.
+	pragmaOptions := []struct {
+		name  string
+		value string
+	}{
+		{
+			name: "busy_timeout",
+			value: fmt.Sprintf(
+				"%d", cfg.BusyTimeout.Milliseconds(),
+			),
+		},
+		{
+			name:  "foreign_keys",
+			value: "on",
+		},
+		{
+			name:  "journal_mode",
+			value: "WAL",
+		},
+	}
+	sqliteOptions := make(url.Values)
+	for _, option := range pragmaOptions {
+		sqliteOptions.Add(
+			sqliteOptionPrefix,
+			fmt.Sprintf("%v=%v", option.name, option.value),
+		)
+	}
+
+	// Then we add any user specified pragma options.
+	for _, option := range cfg.PragmaOptions {
+		pair := strings.Split(option, "=")
+		if len(pair) != 2 {
+			return nil, fmt.Errorf("invalid pragma option: %v. "+
+				"Expected the format: name=value", option)
+		}
+
+		sqliteOptions.Add(
+			sqliteOptionPrefix,
+			fmt.Sprintf("%v=%v", pair[0], pair[1]),
+		)
+	}
+
+	// Construct the DSN which is just the database file name, appended
+	// with the series of pragma options as a query URL string. For more
+	// details on the formatting here, see the modernc.org/sqlite docs:
+	// https://pkg.go.dev/modernc.org/sqlite#Driver.Open.
+	writeDSN := fmt.Sprintf(
+		"%v?%v&%v", filepath.Join(dbPath, fileName),
+		sqliteOptions.Encode(), sqliteTxLockImmediate,
+	)
+	writeCfg := &sqlbase.Config{
+		DriverName:      "sqlite",
+		Dsn:             writeDSN,
+		Timeout:         cfg.Timeout,
+		TableNamePrefix: prefix,
+	}
+
+	writeDB, err := sqlbase.NewSqlBackend(ctx, writeCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	readDSN := fmt.Sprintf(
+		"%v?%v", filepath.Join(dbPath, fileName),
+		sqliteOptions.Encode(),
+	)
+	readCfg := &sqlbase.Config{
+		DriverName:      "sqlite",
+		Dsn:             readDSN,
+		Timeout:         cfg.Timeout,
+		TableNamePrefix: prefix,
+	}
+
+	readDB, err := sqlbase.NewSqlBackend(ctx, readCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sqliteDB{
+		readDBConn:  readDB,
+		writeDBConn: writeDB,
+	}, nil
+}
+
+// sqliteDB contains two sqlite connections, one is optimised for reads as it
+// uses the default BEGIN DEFERRED directive for starting a db transaction and
+// the other is optimised for writes as it uses the BEGIN IMMEDIATE directive
```

Comment:
Perhaps explain why BEGIN IMMEDIATE optimizes for writes?

I saw https://github.com/lightningnetwork/lnd/pull/7252#issuecomment-1361685580 and that DEFERRED creates busy errors. It seems that IMMEDIATE makes this less likely to happen, but is it true that it still can happen when a write tx runs for longer than the busy timeout while another write is waiting for that duration?

Reply:
So immediate will immediately grab the exclusive lock with means no other transactions can start. Without this, it's possible that with deferred, a transaction starts as a read, then needs to upgrade to a write and gets locked out, which triggers the busy timeout. With this we're trying to mark all write transactions as such early, so it doesn't need to upgrade at all. 

> An EXCLUSIVE lock is needed in order to write to the database file. Only one EXCLUSIVE lock is allowed on the file and no other locks of any kind are allowed to coexist with an EXCLUSIVE lock. In order to maximize concurrency, SQLite works to minimize the amount of time that EXCLUSIVE locks are held.

It's still possible for this to return `sqlite_busy` though if another write is started on _another db connection_: 
> IMMEDIATE cause the database connection to start a new write immediately, without waiting for a write statement. The BEGIN IMMEDIATE might fail with [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) if another write transaction is already active on another database connection.

As is, we have a distinct DB connection for writes, so this doesn't appear to apply. 


If it's possible instead have the library let us set the txn type on the single connection level, then we'd just switch over to that, and we wouldn't need to worry about having multiple db connections.

Reply:
>it's possible that with deferred, a transaction starts as a read, then needs to upgrade to a write and gets locked out, which triggers the busy timeout.

Thinking about this more - if there is one dedicated connection for reads and one for writes, how can a write get locked out really? That one write connection can only handle one write tx, so there can never be another one? I am probably still missing a part of the model, perhaps related to the [threading model](https://www.sqlite.org/threadsafe.html) that sqlite runs in?

>(also note that this is part of the other PR, we need to merge that and tag in order to move forward)

I thought the code is introduced in this PR, so I comment on it here?

Reply:
> Thinking about this more - if there is one dedicated connection for reads and one for writes ("As is, we have a distinct DB connection for writes, so this doesn't appear to apply."), how can a write get locked out really?

The reason we added the two connections, was so we can use `BEGIN IMMEDIATE` via a round about way. We're trying to allow the [following situation to be possible](https://www.sqlite.org/isolation.html): 
> If X starts a transaction that will initially only read but X knows it will eventually want to write and does not want to be troubled with possible SQLITE_BUSY_SNAPSHOT errors that arise because another connection jumped ahead of it in line, then X can issue [BEGIN IMMEDIATE](https://www.sqlite.org/lang_transaction.html#immediate) to start its transaction instead of just an ordinary BEGIN. The [BEGIN IMMEDIATE](https://www.sqlite.org/lang_transaction.html#immediate) command goes ahead and starts a write transaction, and thus blocks all other writers. If the [BEGIN IMMEDIATE](https://www.sqlite.org/lang_transaction.html#immediate) operation succeeds, then no subsequent operations in that transaction will ever fail with an [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) error.

A write can get locked out simply because another writer is active. A reader can get kicked out (`SQLITE_BUSY`) if it starts a read, then attempts an update while another write transaction exists concurrently that's also updating. 

>That one write connection can only handle one write tx, so there can never be another one?

Yeah there can only be a single write transaction active for a connection, where "connection" is defined as the object returned from `sqlite.Open`. While that is active, the others will block iiuc. 

Also I got confused myself w.r.t which PR I was commenting on 😅

Reply:
> review and then splitting up right before merge to get the tagging right wouldn't have been a bad option either.

the other PR does contain all the commits. So could just comment on there?

Reply:
Not sure if I am way off here, but I have opened an issue on the modernc.org/sqlite repo to see if they cant set/not set the begin directives based on the passed in driver.TxOptions: https://gitlab.com/cznic/sqlite/-/issues/127

If im on the right track there then that would remove the need for the two conns on our end

Reply:
>the other PR does contain all the commits. So could just comment on there?

Yes, probably should have done that from the start. But then I didn't understand how they were related. And now, this is where we are 😄

I think that issue that you opened is a really smart solution to the problem.

Reply:
> I think that issue that you opened is a really smart solution to the problem.

awesome! I have opened a PR on their repo. Hopefully we can have fast turn around on that side. Does look like they create a new tag after almost every PR they merge.

Reply:
The MR on the gitlab sqlite repo was merged and a new tag was released 🎉  so I have removed this two-connection work around now